### PR TITLE
OIDC: extract success/failure handlers from token login function

### PR DESCRIPTION
**Pull request by kerryarchibald**
For https://github.com/vector-im/element-web/issues/25657
Extracts success/failure handlers from `Lifecycle.attemptTokenLogin` function so they can be reused post OIDC authorization.

## Checklist

-   [ ] Tests written for new code (and old code if feasible)
-   [ ] Linter and other CI checks pass
-   [ ] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

### CODE REVIEW

**Comment by kerryarchibald** [comment]
Strict error is not related to these changes.

File: src/Lifecycle.ts [hunk 1]
```diff
@@ -219,40 +226,62 @@ export function attemptTokenLogin(
...
-            logger.error(err);
+                tryAgainCallback,
+            );
+            logger.error("Failed to log in with login token:", error);
             return false;
         });
 }
 
+/**
+ * After a successful token login or OIDC authorization
```
Comment (richvdh) [inline src/Lifecycle.ts:258]: Clarity nit:

```suggestion
 * Called after a successful token login or OIDC authorization.
```

File: src/Lifecycle.ts [hunk 2]
```diff
@@ -219,40 +226,62 @@ export function attemptTokenLogin(
...
+ * Display a friendly error to the user when token login or OIDC authorization fails
+ * @param description error description
+ * @param tryAgain OPTIONAL function to call on try again button from error dialog
+ */
+async function onFailedDelegatedAuthLogin(description: string | ReactNode, tryAgain?: TryAgainFunction): Promise<void> {
+    Modal.createDialog(ErrorDialog, {
+        title: _t("We couldn't log you in"),
+        description,
+        button: _t("Try again"),
+        onFinished: tryAgain,
```
Comment (richvdh) [inline src/Lifecycle.ts:281]: why do we pass `shouldTryAgain` into the callback function? wouldn't it make more sense for this to be

```suggestion
        onFinished: (shouldTryAgain) => { if (shouldTryAgain && !!tryAgain ) { tryAgain(); } },
```

or something to that effect?

File: src/Lifecycle.ts [hunk 3]
```diff
@@ -204,6 +205,12 @@ export function attemptTokenLogin(
     const identityServer = localStorage.getItem(SSO_ID_SERVER_URL_KEY) ?? undefined;
     if (!homeserver) {
         logger.warn("Cannot log in with token: can't determine HS URL to use");
+        onFailedDelegatedAuthLogin(
+            _t(
+                "We asked the browser to remember which homeserver you use to let you sign in, " +
+                    "but unfortunately your browser has forgotten it. Go to the sign in page and try again.",
+            ),
+        );
         Modal.createDialog(ErrorDialog, {
```
Comment (richvdh) [inline src/Lifecycle.ts:214]: this dialog creation looks redundant now?

**Review by richvdh — CHANGES_REQUESTED** [review]

File: src/Lifecycle.ts (hunk 2 above)
Comment (kerryarchibald) [reply src/Lifecycle.ts:281]: The dialog handles the case of a falsy `onFinished`

Comment (richvdh) [reply src/Lifecycle.ts:273]: I'm not following, I'm afraid.

File: src/Lifecycle.ts [hunk 4]
```diff
@@ -204,55 +205,75 @@ export function attemptTokenLogin(
...
+            const tryAgainCallback: TryAgainFunction = (shouldTryAgain) => {
+                if (shouldTryAgain) {
+                    const cli = createClient({
+                        baseUrl: homeserver,
+                        idBaseUrl: identityServer,
+                    });
+                    const idpId = localStorage.getItem(SSO_IDP_ID_KEY) || undefined;
+                    PlatformPeg.get()?.startSingleSignOn(cli, "sso", fragmentAfterLogin, idpId, SSOAction.LOGIN);
+                }
+            };
```
Comment (richvdh) [inline src/Lifecycle.ts:236]: Per the other thread: I don't understand why the tryAgainCallback is called even if we don't actually want to try again.

```suggestion
            const tryAgainCallback: TryAgainFunction = () => {
                const cli = createClient({
                    baseUrl: homeserver,
                    idBaseUrl: identityServer,
                });
                const idpId = localStorage.getItem(SSO_IDP_ID_KEY) || undefined;
                PlatformPeg.get()?.startSingleSignOn(cli, "sso", fragmentAfterLogin, idpId, SSOAction.LOGIN);
            };
```

File: src/Lifecycle.ts (hunk 2 above)
Comment (richvdh) [reply src/Lifecycle.ts:273]: oh, now I think I see what you mean (if `onFinished` is falsey, it won't be called), but that's not really my point here.

My point is that I don't think it should be up to the `tryAgain` callback to check whether the user pressed "Try again" or "Cancel".

**Review by richvdh — APPROVED** [review]
lgtm on condition of pulling the `if (shouldTryAgain)` condition up to `onFailedDelegatedAuthLogin`

**Pull request merged**